contracts/protocol/integration/amm/ArrakisUniswapV3AmmAdapter.sol

```diff
@@ -0,0 +1,250 @@
+    {   
+        address setToken = _setToken;
+        uint256[] memory minTokensOut = _minTokensOut;
+        uint256 liquidity = _liquidity;
+        IArrakisVaultV1 arrakisVaultPool = IArrakisVaultV1(_pool);
+
+        // Make sure that only up to the amount of liquidity tokens owned by the Set Token are redeemed
+        uint256 setTokenLiquidityBalance = arrakisVaultPool.balanceOf(setToken);
+        require(liquidity <= setTokenLiquidityBalance, "_liquidity must be <= to current balance");
+
+        // Checks for minTokensOut required?
+
+        target = router;
+        value = 0;
+        data = abi.encodeWithSignature(
+            REMOVE_LIQUIDITY,
+            arrakisVaultPool,
+            liquidity,
+            minTokensOut[0],
+            minTokensOut[1],
+            setToken
+        );
+    }
+
+    /**
+     * Return calldata for the remove liquidity single asset call
+     */
+    function getRemoveLiquiditySingleAssetCalldata(
+        address /* _setToken */,
+        address /*_pool*/,
+        address /*_component*/,
+        uint256 /*_minTokenOut*/,
+        uint256 /*_liquidity*/
+    )
+        external
+        view
+        override
+        returns (address /*target*/, uint256 /*value*/, bytes memory /*data*/)
+    {
+        revert("Arrakis single asset removal is not supported");
+    }
+
+    /**
+     * Returns the address of the spender
+     */
+    function getSpenderAddress(address /*_pool*/)
+        external
+        view
+        override
+        returns (address spender)
+    {
+        spender = router;
+    }
+
+    /**
+     * Verifies that this is an Arrakis Vault pool holding valid UniswapV3 position
+     *
+     * @param  _pool          Address of liquidity token
+     * @param  _components    Address array of supplied/requested tokens
+     */
+    function isValidPool(address _pool, address[] memory _components)
+        external
+        view
+        override
+        returns (bool)
+    {
+        // Attempt to get the tokens of the provided pool
+        address token0;
+        address token1;
+        try IArrakisVaultV1(_pool).token0() returns (IERC20 _token0) {
+            token0 = address(_token0);
+        } catch {
+            return false;
+        }
+        try IArrakisVaultV1(_pool).token1() returns (IERC20 _token1) {
+            token1 = address(_token1);
+        } catch {
+            return false;
+        }
+
+        // Make sure the tokens stored in the arrakis pools matches the
+        // tokens provided in the components
+        if (
+            _components.length != 2 ||
+            !(token0 == _components[0] || token0 == _components[1]) ||
+            !(token1 == _components[0] || token1 == _components[1])
+        ) {
+            return false;
+        }
+
+        // Make sure the pool address follows IERC20 interface
+        try IArrakisVaultV1(_pool).totalSupply() returns (uint256 _totalSupply) {
```

Comment:
I think we should just check that _totalSupply > 0

Reply:
actually, _totalSupply can be zero if the vault has been just deployed and we're adding liquidity for the first time. 
_totalSupply increases when we add liquidity and it mint tokens to us.